jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/params/TreeBoosterParams.scala

```diff
@@ -0,0 +1,238 @@
+  final val processType = new Param[String](this, "process_type", "A type of boosting process to " +
+    "run. options: {default, update}",
+    ParamValidators.inArray(Array("default", "update")))
+
+  final def getProcessType: String = $(processType)
+
+  final val growPolicy = new Param[String](this, "grow_policy", "Controls a way new nodes are " +
+    "added to the tree. Currently supported only if tree_method is set to hist or approx. " +
+    "Choices: depthwise, lossguide. depthwise: split at nodes closest to the root. " +
+    "lossguide: split at nodes with highest loss change.",
+    ParamValidators.inArray(Array("depthwise", "lossguide")))
+
+  final def getGrowPolicy: String = $(growPolicy)
+
+
+  final val maxLeaves = new IntParam(this, "max_leaves", "Maximum number of nodes to be added. " +
+    "Not used by exact tree method", ParamValidators.gtEq(0))
+
+  final def getMaxLeaves: Int = $(maxLeaves)
+
+  final val maxBins = new IntParam(this, "max_bin", "Maximum number of discrete bins to bucket " +
+    "continuous features. Increasing this number improves the optimality of splits at the cost " +
+    "of higher computation time. Only used if tree_method is set to hist or approx.",
+    ParamValidators.gt(0))
+
+  final def getMaxBins: Int = $(maxBins)
+
+  final val numParallelTree = new IntParam(this, "num_parallel_tree", "Number of parallel trees " +
+    "constructed during each iteration. This option is used to support boosted random forest.",
+    ParamValidators.gt(0))
+
+  final def getNumParallelTree: Int = $(numParallelTree)
+
+  final val monotoneConstraints = new IntArrayParam(this, "monotone_constraints", "Constraint of " +
+    "variable monotonicity.")
+
+  final def getMonotoneConstraints: Array[Int] = $(monotoneConstraints)
+
+  final val interactionConstraints = new Param[String](this,
+    name = "interaction_constraints",
+    doc = "Constraints for interaction representing permitted interactions. The constraints" +
+      " must be specified in the form of a nest list, e.g. [[0, 1], [2, 3, 4]]," +
+      " where each inner list is a group of indices of features that are allowed to interact" +
+      " with each other. See tutorial for more information")
+
+  final def getInteractionConstraints: String = $(interactionConstraints)
+
+
+  final val maxCachedHistNode = new IntParam(this, "max_cached_hist_node", "Maximum number of " +
+    "cached nodes for CPU histogram.",
+    ParamValidators.gt(0))
+
+  final def getMaxCachedHistNode: Int = $(maxCachedHistNode)
+
+  setDefault(eta -> 0.3, gamma -> 0, maxDepth -> 6, minChildWeight -> 1, maxDeltaStep -> 0,
+    subsample -> 1, samplingMethod -> "uniform", colsampleBytree -> 1, colsampleBylevel -> 1,
+    colsampleBynode -> 1, lambda -> 1, alpha -> 0, treeMethod -> "auto", scalePosWeight -> 1,
+    processType -> "default", growPolicy -> "depthwise", maxLeaves -> 0, maxBins -> 256,
+    numParallelTree -> 1, maxCachedHistNode -> 65536)
+
```

Comment:
Is it necessary to copy the default here? Is it possible to have an Optional type that indicates not-set and let xgboost handle the default?

Reply:
This is just default value and if user doesn't set them explicitly, they will not be passed to xgboost.

Reply:
So, we can set arbitrary values here?

Reply:
Yeah, that's true. But those values are the defaults from xgboost python parameters page.

Reply:
Yeah, we need to "remember" changing this whenever we change the internal, but we always forget, hence the outdated parameter list.

Reply:
If we don't set the default values, the getter function will throw exception. That's really painful.